File: leetcode/Python/Lc2.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        ans = ListNode(0)
        p = ans
        sum1 = 0
        sum2 = 0
        b = 0
        p1 = l1
        while p1 != None:
            sum1 += p1.val * (10 ** b)
            b += 1
            p1 = p1.next

        b = 0

        p2 = l2
        while p2 != None:
            sum2 += p2.val * (10 ** b)
            b += 1
            p2 = p2.next

        sum = sum1 + sum2
        i = 0
        while sum/(10**i)>=1:
            i += 1

        if sum == 0:
            i += 1

        j  =0
        while j<i:
            p.next = ListNode((sum//(10**j))%10)
            p = p.next
            j += 1

        ans = ans.next
        return ans

l1 = ListNode(2)
l1.next = ListNode(4)
l1.next.next = ListNode(3)
l2 = ListNode(5)
l2.next = ListNode(6)
l2.next.next = ListNode(4)

ans = Solution().addTwoNumbers(l1,l2)

File: leetcode/Java/src/Lc2.java
/*
the problem is not solved in Java
please look the Python version
 */

public class Lc2 {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode dummyHead = new ListNode(0);
        ListNode p = l1, q = l2, curr = dummyHead;
        int carry = 0;
        while (p != null || q != null) {
            int x = (p != null) ? p.val : 0;
            int y = (q != null) ? q.val : 0;
            int sum = carry + x + y;
            carry = sum / 10;
            curr.next = new ListNode(sum % 10);
            curr = curr.next;
            if (p != null) p = p.next;
            if (q != null) q = q.next;
        }
        if (carry > 0) {
            curr.next = new ListNode(carry);
        }
        return dummyHead.next;
    }

    public static void main(String[] args) {
        ListNode l1 = new ListNode(9);
        ListNode l2 = new ListNode(1);
        ListNode p = l2;
        for (int i=0; i<9; i++) {
            p.next = new ListNode(9);
            p = p.next;
        }
        new Lc2().addTwoNumbers(l1,l2);
    }
}


class ListNode {
    int val;
    ListNode next;
    ListNode(int x) { val = x; }
}
